modules/algorithm/puzzle_15_bfs.py

```diff
@@ -0,0 +1,75 @@
+class State:
+    def __init__(self, initial_state):
+        self.initial_state = tuple(initial_state)
+
+    def __eq__(self, other):
+        return self.initial_state == other.initial_state
+    
+    def __hash__(self):
+        return hash(tuple(self.initial_state))
+    
+        
+    def generate_neighbours(self):
+        first_state = self.initial_state
+        neighbour = []
+
+        moves = {
+            'up':-3,
+            'down':+3,
+            'left':-1,
+            'right':+1
+        }
+        invalid_left = [0,3,6]
+        invalid_right = [2,5,8]
+
+        #finding the first instance of 0
+        index = first_state.index(0)
+
+
+        for direction, v in moves.items():
+            new_index = index + v
+            if direction == 'left' and index in invalid_left:
+                continue
+            if direction == 'right' and index in invalid_right:
+                continue
+            if 0 <= new_index < 9:
+                new_initial_state = list(first_state) # converting to list to modify
+                new_initial_state[index],new_initial_state[new_index] =  new_initial_state[new_index], new_initial_state[index]
+                neighbour.append(tuple(new_initial_state)) #making it hashable
```

Comment:
You don’t need to cast it as a tuple here (because you do this in the constructor!)